<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Iterative Correcting Weighted Gene Co-expression Network Analysis function for constructing a gene network from a gene expression matrix. The algorithm: 1. Constructs a signed wgcna network. 2. Drops correlated modules based on kurtosis. 3. Regresses out the largest community from the expression data. 4. Repeats steps 1-3 until a maximum number of communities or iterations is reached.">
<title>Iterative Correcting Weighted Gene Co-expression Network Analysis â€¢ icWGCNA</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Iterative Correcting Weighted Gene Co-expression Network Analysis">
<meta property="og:description" content="Iterative Correcting Weighted Gene Co-expression Network Analysis function for constructing a gene network from a gene expression matrix. The algorithm: 1. Constructs a signed wgcna network. 2. Drops correlated modules based on kurtosis. 3. Regresses out the largest community from the expression data. 4. Repeats steps 1-3 until a maximum number of communities or iterations is reached.">
<meta name="google-site-verification" content="iibkm1sdjFL0gSzOkeENIdmk9u5Jkc6ZFQc4q-vTSpQ">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Google tag (gtag.js) --><script async src="https://www.googletagmanager.com/gtag/js?id=G-1ZQ1CR751E"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-1ZQ1CR751E');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">icWGCNA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.5.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Systems-Methods/icWGCNA/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="icwgcna">icWGCNA<a class="anchor" aria-label="anchor" href="#icwgcna"></a>
</h1></div>
<!-- badges: start -->

<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>Iterative Correcting Weighted Gene Co-expression Network Analysis function for constructing a gene network from a gene expression matrix. The algorithm:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fl">1.</span> Constructs a signed wgcna network</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fl">2.</span> Drops correlated modules based on membership kurtosis</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fl">3.</span> Regresses out the strongest community from the expression data</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="fl">4.</span> Repeats steps <span class="dv">1-3</span> until a maximum number of communities or iterations is reached</span></code></pre></div>
<p>Some differences from standard <a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-9-559" class="external-link">WGNCA (Horvath/Langfelder)</a>:</p>
<ul>
<li>Makes heavy use of <a href="https://cran.r-project.org/web/packages/Rfast/" class="external-link">Rfast</a> to compute adjacencies and <a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-9-559" class="external-link">topological overlap measure (TOM)</a>, which enables iterative network creation on &gt; 20K features.</li>
<li>Always uses <a href="https://bmcgenomics.biomedcentral.com/articles/10.1186/1471-2164-10-327" class="external-link">signed adjacency</a> in order to avoid possible distortions of community signatures (eigengenes).</li>
<li>Iteratively regresses out strongest community in order to facilitate discovery of communities possibly obscured by larger module(s).</li>
<li>Clustering does not focus on merging communities but dropping to identify strongest module(s), keeping communities with higher membership kurtosis.</li>
<li>In addition to Pearson correlation, there is an option for Spearman correlation as the base constructing adjacency matrix measure in order to enable robust application in RNA-seq and micro-array data. Future updates may include mutual information and other measures of similarity.</li>
</ul>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>Install the released version of icWGCNA from Github:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span>repo <span class="op">=</span> <span class="st">"Systems-Methods/icWGCNA"</span><span class="op">)</span></span></code></pre></div>
<p>Or install the development version from BMS BioGit with:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span>repo <span class="op">=</span> <span class="st">"Systems-Methods/icWGCNA"</span>, </span>
<span>                        ref <span class="op">=</span> <span class="st">"develop"</span><span class="op">)</span></span></code></pre></div>
<p>Or:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_git</span><span class="op">(</span></span>
<span>  <span class="st">'https://github.com/Systems-Methods/icWGCNA.git'</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="example">Example<a class="anchor" aria-label="anchor" href="#example"></a>
</h2>
<div class="section level3">
<h3 id="example-data">Example Data<a class="anchor" aria-label="anchor" href="#example-data"></a>
</h3>
<p>First we can use the <a href="https://cran.r-project.org/web/packages/UCSCXenaTools/vignettes/USCSXenaTools.html" class="external-link">{UCSCXenaTools}</a> package to download an example mRNASeq dataset.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">luad</span> <span class="op">&lt;-</span> <span class="fu">UCSCXenaTools</span><span class="fu">::</span><span class="fu"><a href="https://docs.ropensci.org/UCSCXenaTools/reference/getTCGAdata.html" class="external-link">getTCGAdata</a></span><span class="op">(</span>project <span class="op">=</span> <span class="st">"LUAD"</span>, </span>
<span>                                   mRNASeq <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                   mRNASeqType <span class="op">=</span> <span class="st">"normalized"</span>,</span>
<span>                                   clinical <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                                   download <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ex</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="va">luad</span><span class="op">$</span><span class="va">destfiles</span><span class="op">)</span>, rownames <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="main-results">Main Results<a class="anchor" aria-label="anchor" href="#main-results"></a>
</h3>
<p>Next we can run the main <code><a href="reference/icwgcna.html">icwgcna()</a></code> function</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Systems-Methods/icWGCNA" class="external-link">icWGCNA</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/icwgcna.html">icwgcna</a></span><span class="op">(</span><span class="va">ex</span>,   </span>
<span>                   expo <span class="op">=</span> <span class="fl">6</span>,</span>
<span>                   Method <span class="op">=</span> <span class="st">"pearson"</span>,</span>
<span>                   q <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>                   maxIt <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                   maxComm <span class="op">=</span> <span class="fl">100</span>,</span>
<span>                   corCut <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>                   covCut <span class="op">=</span> <span class="fl">0.33</span>,</span>
<span>                   mat_mult_method <span class="op">=</span> <span class="st">"Rfast"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="downstream-analysis">Downstream Analysis<a class="anchor" aria-label="anchor" href="#downstream-analysis"></a>
</h3>
<p>Finally, downstream analysis can be run on the Iterative Correcting Weighted Gene Co-expression Network Analysis results.</p>
<div class="section level4">
<h4 id="compute-community-signatures-eigengenes">Compute community signatures (eigengenes)<a class="anchor" aria-label="anchor" href="#compute-community-signatures-eigengenes"></a>
</h4>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">eigengene_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/compute_eigengene_matrix.html">compute_eigengene_matrix</a></span><span class="op">(</span></span>
<span>  <span class="va">ex</span>, </span>
<span>  membership_matrix <span class="op">=</span> <span class="va">results</span><span class="op">$</span><span class="va">community_membership</span>, </span>
<span>  cutoff <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  pc_flag <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="compute-panglaodb-collection-enrichments-for-each-community">Compute panglaoDB collection enrichments for each community<a class="anchor" aria-label="anchor" href="#compute-panglaodb-collection-enrichments-for-each-community"></a>
</h4>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pangDB</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="va">pangDB_link</span><span class="op">)</span></span>
<span><span class="va">panglaoDB_enrichment</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/compute_panglaoDB_enrichment.html">compute_panglaoDB_enrichment</a></span><span class="op">(</span></span>
<span>  <span class="va">results</span><span class="op">$</span><span class="va">community_membership</span>,</span>
<span>  K <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  memb_cut <span class="op">=</span> <span class="fl">0.65</span>,</span>
<span>  pangDB <span class="op">=</span> <span class="va">pangDB</span>,</span>
<span>  prolif <span class="op">=</span> <span class="va">prolif_names</span>, </span>
<span>  p_cut <span class="op">=</span> <span class="fl">0.001</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="compute-msigdb-collection-enrichments-for-each-community">Compute MSigDB collection enrichments for each community<a class="anchor" aria-label="anchor" href="#compute-msigdb-collection-enrichments-for-each-community"></a>
</h4>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MSigDB_enrichment</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/compute_MSigDB_enrichment.html">compute_MSigDB_enrichment</a></span><span class="op">(</span></span>
<span>  <span class="va">results</span><span class="op">$</span><span class="va">community_membership</span>,</span>
<span>  K <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  memb_cut <span class="op">=</span> <span class="fl">.65</span>,</span>
<span>  cats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"H"</span>, <span class="st">"C3"</span>, <span class="st">"C6"</span>, <span class="st">"C7"</span>, <span class="st">"C8"</span><span class="op">)</span>, </span>
<span>  p_cut <span class="op">=</span> <span class="fl">0.001</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="compute-xcell-collection-enrichments-for-each-community">Compute xCell collection enrichments for each community<a class="anchor" aria-label="anchor" href="#compute-xcell-collection-enrichments-for-each-community"></a>
</h4>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xCell_enrichment</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/compute_xCell_enrichment.html">compute_xCell_enrichment</a></span><span class="op">(</span></span>
<span>  <span class="va">results</span><span class="op">$</span><span class="va">community_membership</span>,</span>
<span>  K <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  memb_cut <span class="op">=</span> <span class="fl">.65</span>, </span>
<span>  p_cut <span class="op">=</span> <span class="fl">0.001</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="display-umap-of-community-membership-with-text-overlays">Display UMAP of Community Membership with text overlays<a class="anchor" aria-label="anchor" href="#display-umap-of-community-membership-with-text-overlays"></a>
</h4>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">network_umap</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/make_network_umap.html">make_network_umap</a></span><span class="op">(</span></span>
<span>  <span class="va">results</span><span class="op">$</span><span class="va">community_membership</span>,</span>
<span>  community_memb_cut_main <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>  community_n_main <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  community_memb_cut_secondary <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  community_n_secondary <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  gene_memb_cut_main <span class="op">=</span> <span class="fl">0.75</span>,</span>
<span>  gene_memb_cut_secondary <span class="op">=</span> <span class="fl">0.65</span>,</span>
<span>  community_labels <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  umap_specs <span class="op">=</span> <span class="fu">umap</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/umap/man/umap.defaults.html" class="external-link">umap.defaults</a></span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">network_umap</span><span class="op">$</span><span class="va">umap_w_annotation</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="identify-top-gene-of-communities-that-are-unique-only-belong-to-one-community">Identify Top Gene of Communities that are unique (only belong to one community)<a class="anchor" aria-label="anchor" href="#identify-top-gene-of-communities-that-are-unique-only-belong-to-one-community"></a>
</h4>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">unique_top_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/find_unique_top_genes.html">find_unique_top_genes</a></span><span class="op">(</span></span>
<span>  <span class="va">results</span><span class="op">$</span><span class="va">community_membership</span>,</span>
<span>  K <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  maxIt <span class="op">=</span> <span class="fl">10</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="code-of-conduct">Code of Conduct<a class="anchor" aria-label="anchor" href="#code-of-conduct"></a>
</h2>
<p>Please note that the icWGCNA project is released with a <a href="https://contributor-covenant.org/version/2/0/CODE_OF_CONDUCT.html" class="external-link">Contributor Code of Conduct</a>. By contributing to this project, you agree to abide by its terms.</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/Systems-Methods/icWGCNA/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/Systems-Methods/icWGCNA/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>

<div class="community">
<h2 data-toc-skip>Community</h2>
<ul class="list-unstyled">
<li><a href="CODE_OF_CONDUCT.html">Code of conduct</a></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing icWGCNA</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="https://github.com/mjrmason" class="external-link">Mike Mason</a> <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-5652-7739" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>
<a href="https://github.com/wfulp" class="external-link">Jimmy Fulp</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0001-6888-0794" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://lifecycle.r-lib.org/articles/stages.html#experimental" class="external-link"><img src="https://img.shields.io/badge/lifecycle-experimental-orange.svg" alt="Lifecycle: experimental"></a></li>
<li><a href="https://app.codecov.io/gh/Systems-Methods/icWGCNA?branch=main" class="external-link"><img src="https://codecov.io/gh/Systems-Methods/icWGCNA/branch/main/graph/badge.svg" alt="Codecov test coverage"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/mjrmason" class="external-link">Mike Mason</a>, <a href="https://github.com/wfulp" class="external-link">Jimmy Fulp</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
