<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Compute community signatures (eigengenes) — compute_eigengene_matrix • icWGCNA</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Compute community signatures (eigengenes) — compute_eigengene_matrix"><meta name="description" content="Computes the community signatures (eigengenes) for an expression matrix
given a particular community membership (kME) matrix"><meta property="og:description" content="Computes the community signatures (eigengenes) for an expression matrix
given a particular community membership (kME) matrix"><meta name="google-site-verification" content="iibkm1sdjFL0gSzOkeENIdmk9u5Jkc6ZFQc4q-vTSpQ"><!-- Google tag (gtag.js) --><script async src="https://www.googletagmanager.com/gtag/js?id=G-1ZQ1CR751E"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-1ZQ1CR751E');
</script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">icWGCNA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Systems-Methods/icWGCNA/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Compute community signatures (eigengenes)</h1>
      <small class="dont-index">Source: <a href="https://github.com/Systems-Methods/icWGCNA/blob/main/R/downstream.R" class="external-link"><code>R/downstream.R</code></a></small>
      <div class="d-none name"><code>compute_eigengene_matrix.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Computes the community signatures (eigengenes) for an expression matrix
given a particular community membership (kME) matrix</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">compute_eigengene_matrix</span><span class="op">(</span><span class="va">ex</span>, <span class="va">membership_matrix</span>, cutoff <span class="op">=</span> <span class="fl">5</span>, pc_flag <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-ex">ex<a class="anchor" aria-label="anchor" href="#arg-ex"></a></dt>
<dd><p>matrix of bulk RNA-seq or microarray gene expression data.
This should be in log space and greater than or equal to 0.</p></dd>


<dt id="arg-membership-matrix">membership_matrix<a class="anchor" aria-label="anchor" href="#arg-membership-matrix"></a></dt>
<dd><p>a community membership (kME) matrix with genes as
rows and communities as columns. Often <code>community_membership</code> or
<code>full_community_membership</code> output from <code><a href="icwgcna.html">icwgcna()</a></code></p></dd>


<dt id="arg-cutoff">cutoff<a class="anchor" aria-label="anchor" href="#arg-cutoff"></a></dt>
<dd><p>number of top genes to use when computing community signatures</p></dd>


<dt id="arg-pc-flag">pc_flag<a class="anchor" aria-label="anchor" href="#arg-pc-flag"></a></dt>
<dd><p>indicator. TRUE (default) means to use the 1st principal component
(corrected for direction). FALSE uses the mean of scaled and centered top genes.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A matrix with rows being the community signature and columns being samples</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This can be used to compute community signatures in a new expression dataset.
Note, community signatures are not corrected by <code><a href="icwgcna.html">icwgcna()</a></code>
iterations so it will not match signatures output from <code><a href="icwgcna.html">icwgcna()</a></code>
if it is run on the network construction dataset.</p>
<p>When using these community signatures for modeling it may be best to include
interaction terms or use tree based
methods since dependencies are not addressed in this output matrix.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://docs.ropensci.org/UCSCXenaTools/" class="external-link">"UCSCXenaTools"</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">luad</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://docs.ropensci.org/UCSCXenaTools/reference/getTCGAdata.html" class="external-link">getTCGAdata</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  project <span class="op">=</span> <span class="st">"LUAD"</span>, mRNASeq <span class="op">=</span> <span class="cn">TRUE</span>, mRNASeqType <span class="op">=</span> <span class="st">"normalized"</span>,</span></span>
<span class="r-in"><span>  clinical <span class="op">=</span> <span class="cn">FALSE</span>, download <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">ex</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="va">luad</span><span class="op">$</span><span class="va">destfiles</span><span class="op">)</span>, rownames <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="icwgcna.html">icwgcna</a></span><span class="op">(</span><span class="va">ex</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">compute_eigengene_matrix</span><span class="op">(</span><span class="va">ex</span>, <span class="va">results</span><span class="op">$</span><span class="va">community_membership</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/mjrmason" class="external-link">Mike Mason</a>, <a href="https://github.com/wfulp" class="external-link">Jimmy Fulp</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

