<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Iterative Correcting Weighted Gene Co-expression Network Analysis function constructing a network from an expression matrix."><title>icwgcna — icwgcna • icWGCNA</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="icwgcna — icwgcna"><meta property="og:description" content="Iterative Correcting Weighted Gene Co-expression Network Analysis function constructing a network from an expression matrix."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">icWGCNA</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Unreleased version">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>icwgcna</h1>
      
      <div class="d-none name"><code>icwgcna.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Iterative Correcting Weighted Gene Co-expression Network Analysis function constructing a network from an expression matrix.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">icwgcna</span><span class="op">(</span>
  <span class="va">ex</span>,
  expo <span class="op">=</span> <span class="fl">6</span>,
  Method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pearson"</span>, <span class="st">"spearman"</span><span class="op">)</span>,
  q <span class="op">=</span> <span class="fl">0.3</span>,
  maxIt <span class="op">=</span> <span class="fl">10</span>,
  maxComm <span class="op">=</span> <span class="fl">100</span>,
  corCut <span class="op">=</span> <span class="fl">0.8</span>,
  covCut <span class="op">=</span> <span class="fl">0.33</span>,
  mat_mult_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Rfast"</span>, <span class="st">"RcppEigen"</span><span class="op">)</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>ex</dt>
<dd><p>matrix of bulk RNA-seq or microarray gene expression data. This should be in log space and greater than 0.</p></dd>
<dt>expo</dt>
<dd><p>exponent to use for soft thresholding. If NULL will use angular distance</p></dd>
<dt>Method</dt>
<dd><p>correlation to use for distance measure, "pearson" (default) or "spearman"</p></dd>
<dt>q</dt>
<dd><p>quantile (0-1) for first round filtering based on mean expression and standard deviation</p></dd>
<dt>maxIt</dt>
<dd><p>maximum number of iterations must be 25 or less</p></dd>
<dt>maxComm</dt>
<dd><p>maximum number of communities to be found</p></dd>
<dt>corCut</dt>
<dd><p>correlation threshold used for dropping communities</p></dd>
<dt>covCut</dt>
<dd><p>coefficient of variation (CoV) quantile threshold to use at each iteration  for selecting genes to build network. covCut = .667 would use the top third of genes based on CoV after regressing out largest community</p></dd>
<dt>mat_mult_method</dt>
<dd><p>method for large matrix multiplication, "Rfast" (default) or "RcppEigen" (see details)</p></dd>
</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Returns a list with the following items:</p><ul><li><p><code>community_membership</code> - community membership score (kME). Analogous to loadings in PCA.</p></li>
<li><p><code>community_signature</code> - community eigengene, the first principal component the expression of genes in this community. This can be thought of as the average of the scaled expression of top community genes.</p></li>
<li><p><code>full_community_membership</code> -  similar to community_membership but includes communities that were dropped in the iterative process.</p></li>
<li><p><code>full_community_signature</code> - similar to community_signature but includes communities that were dropped in the iterative process.</p></li>
<li><p><code>controlled_for</code> - The communities whose signatures were regressed out at each iteration.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Iterative Correcting Weighted Gene Co-expression Network Analysis function for constructing a gene network from a gene expression matrix. The algorithm:</p><ol><li><p>Constructs a signed wgcna network</p></li>
<li><p>Drops correlated modules based on kurtosis.</p></li>
<li><p>Regresses out the largest community from the expression data.</p></li>
<li><p>Repeats steps 1-3 until a maximum number of communities or iterations is reached.</p></li>
</ol><p>Some differences from standard WGNCA (Horvath/Langfelder)</p><ul><li><p>Makes heavy use of <a href="https://rdrr.io/pkg/Rfast/man/Rfast-package.html" class="external-link">Rfast</a> to compute adjacencies and TOM to enable iterative network creation on &gt; 20K features.</p></li>
<li><p>Uses signed adjacency in order to avoid possible distortions of community signatures (eigengenes).</p></li>
<li><p>Iteratively regresses out strongest community in order to facilitate discovery of communities possibly obscured larger module(s).</p></li>
<li><p>Clustering does not focus on merging communities but dropping to identify strongest module(s).</p></li>
<li><p>Enables Spearman correlation for constructing adjacency matrix instead of Pearson to enable robust application in RNA-seq and micro-array data. Future updates may include mutual information</p></li>
</ul><p>For matrix multiplication the option "Rfast" will use <code><a href="https://rdrr.io/pkg/Rfast/man/mat.mult.html" class="external-link">Rfast::mat.mult()</a></code>,
which takes advantage of parallel processing across multiple cores. The option
"RcppEigen" will use the RcppEigen engine for C++ code, which tends to be faster
when using a single core, but does not take advantage of parallel processing across
multiple cores. If running this on a cluster with access to many computer core
there is a significant performance advantage to using <code><a href="https://rdrr.io/pkg/Rfast/man/mat.mult.html" class="external-link">Rfast::mat.mult()</a></code></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"></span>
<span class="r-in"><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span>
<span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://docs.ropensci.org/UCSCXenaTools/" class="external-link">"UCSCXenaTools"</a></span><span class="op">)</span></span>
<span class="r-in"><span class="va">luad</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://docs.ropensci.org/UCSCXenaTools/reference/getTCGAdata.html" class="external-link">getTCGAdata</a></span><span class="op">(</span>project <span class="op">=</span> <span class="st">"LUAD"</span>, mRNASeq <span class="op">=</span> <span class="cn">TRUE</span>, mRNASeqType <span class="op">=</span> <span class="st">"normalized"</span>,</span>
<span class="r-in">                  clinical <span class="op">=</span> <span class="cn">FALSE</span>, download <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span class="r-in"><span class="va">ex</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="va">luad</span><span class="op">$</span><span class="va">destfiles</span><span class="op">)</span></span>
<span class="r-in"><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu">icwgcna</span><span class="op">(</span><span class="va">ex</span><span class="op">)</span><span class="op">}</span></span>
<span class="r-in"></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by <a href="https://biogit.pri.bms.com/masonm3" class="external-link">Mike Mason</a>, <a href="https://biogit.pri.bms.com/fulpw" class="external-link">Jimmy Fulp</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

    </footer></div>

  

  

  </body></html>

